spdx-utils/src/main/kotlin/model/SpdxExternalReference.kt

```diff
@@ -56,25 +63,38 @@ data class SpdxExternalReference(
         OTHER
     }
 
-    // TODO: Using new values as per https://github.com/spdx/spdx-spec/issues/451.
-    enum class Type(
+    sealed class Type(
+        @JsonValue
         val typeName: String,
         val category: Category
     ) {
-        BOWER("bower", Category.PACKAGE_MANAGER),
-        CPE_22_TYPE("cpe22Type", Category.SECURITY),
-        CPE_23_TYPE("cpe23Type", Category.SECURITY),
-        MAVEN_CENTRAL("maven-central", Category.PACKAGE_MANAGER),
-        NPM("npm", Category.PACKAGE_MANAGER),
-        NUGET("nuget", Category.PACKAGE_MANAGER),
-        PURL("purl", Category.PACKAGE_MANAGER),
-        SWH("swh", Category.PERSISTENT_ID),
-        OTHER("OTHER", Category.OTHER);
+        object Cpe22Type : Type("cpe22Type", Category.SECURITY)
+        object Cpe23Type : Type("cpe23Type", Category.SECURITY)
+
+        object Bower : Type("bower", Category.PACKAGE_MANAGER)
+        object MavenCentral : Type("maven-central", Category.PACKAGE_MANAGER)
+        object Npm : Type("npm", Category.PACKAGE_MANAGER)
+        object NuGet : Type("nuget", Category.PACKAGE_MANAGER)
+        object Purl : Type("purl", Category.PACKAGE_MANAGER)
+
+        object SoftwareHeritage : Type("swh", Category.PERSISTENT_ID)
+
+        class Other(typeName: String) : Type(typeName, Category.OTHER)
     }
 
     init {
         require(referenceLocator.isNotBlank()) { "The referenceLocator must not be blank." }
+    }
+}
 
-        require(referenceType.isNotBlank()) { "The referenceType must not be blank." }
+private class ReferenceTypeDeserializer : StdDeserializer<SpdxExternalReference.Type>(
+    SpdxExternalReference.Type::class.java
+) {
+    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): SpdxExternalReference.Type {
+        val node = p.codec.readTree<JsonNode>(p)
+        val typeName = node.textValue()
```

Comment:
Shouldn't you access the `typeName` property of `Type` to get this value?

Reply:
No, that's part of the trick: This deserializer is only for the `referenceType` field. Any while in the YAML / JSON only associates a string value to that field, we construct a more complex / nested type here from that string.